### DISCUSSION TITLE
Model event for deleting files

### DISCUSSION THREAD

**Discussion by DevMushref** [filamentphp/filament#5387] [Showcase]
I've found few discussions here requesting/asking how to delete files from storage. So here is how I did it.

### In your model file
Create a public function called `delete()`.

Within it you have few options to make

**1- If the file is single and NOT required:**

```
if (!is_null($this->attachment) && File::exists('storage/' . $this->attachment)) {
   File::delete('storage/' . $this->statement);
}
```

**2 - If the file is single and IS required:**

```
if (File::exists('storage/' . $this->attachment)) {
   File::delete('storage/' . $this->statement);
}
```

**3- If the files are multiple and are NOT required:**

```
foreach ($this->images as $image) {
     if (!is_null($this->images) && File::exists('storage/' . $image)) {
         File::delete('storage/' . $image);
    }
}
```

**4- If the files are multiple and ARE required:**

```
foreach ($this->images as $image) {
     if ((File::exists('storage/' . $image)) {
         File::delete('storage/' . $image);
    }
}
```

### Why `is_null()`?
When I tried to delete a record that doesn't have multiple images uploaded, I get an error saying that the image directory is NULL.
So what I basically did is tell the software that if the array is NOT null aka `!is_null()` and the files DO exist, then delete them.

### How does the code read the directory URL?
It always saves in the `storage/` folder. In your Filament resource, if you set the directory then it will be part of the `$record` saved in database.

### For example:

I have a `FileUpload` filed that saves files in a folder called `reports`:

```
Forms\Components\FileUpload::make('attachment')
     ->image()
     ->directory('reports')
```

If you want to call the image in a custom blade then you should call it like so:

`<img src="storage/ {{ $record->attachment }}" />`

The URL will be `storage/ + reports/image.png`. If no directory was set: `storage/ + image.png`.

So always keep in mind that Filament saves the directory name (if available) as well as the image name.

Hope it helped in your case!

Thank you for reading.

**Comment by denizaygundev** [comment]
Thanks for sharing @DevMushref 

I found this to be effective:
```
    protected static function booted(): void
    {
        static::deleted(function (File $file) {
            Storage::disk('s3')->delete($file->url);
        });
    }
```

And it will work with one or multiple files (if using `->multiple`)
